ARCs/arc-0020.md

```diff
@@ -0,0 +1,870 @@
+of an Algorand Standard Asset (ASA).
+
+To provide a unified and familiar interface between the Algorand Standard Asset
+and the Smart ASA, method names and parameters have been adapted to the ABI
+types but left otherwise unchanged.
+
+#### Asset Creation
+
+```json
+{
+  "name": "asset_create",
+  "args": [
+    { "type": "uint64", "name": "total" },
+    { "type": "uint32", "name": "decimals" },
+    { "type": "bool", "name": "default_frozen" },
+    { "type": "string", "name": "unit_name" },
+    { "type": "string", "name": "name" },
+    { "type": "string", "name": "url" },
+    { "type": "byte[]", "name": "metadata_hash" },
+    { "type": "address", "name": "manager_addr" },
+    { "type": "address", "name": "reserve_addr" },
+    { "type": "address", "name": "freeze_addr" },
+    { "type": "address", "name": "clawback_addr" }
+  ],
+  "returns": { "type": "uint64" }
+}
+```
+
+Calling `asset_create` creates a new Smart ASA and returns the identifier of the
+ASA. The [metadata section](#metadata) describes its required properties.
+
+> Upon a call to `asset_create`, a reference implementation SHOULD:
+>
+> - Mint an Algorand Standard Asset (ASA) that MUST specify the properties
+>   defined in the [metadata section](#metadata). In addition:
+>   - The `manager`, `reserve` and `freeze` addresses SHOULD be set to the
+>     account of the controlling Smart Contract.
+>   - The remaining fields are left to the implementation, which MAY set `total`
+>     to `2 ** 64 - 1` to enable dynamically increasing the max circulating
+>     supply of the Smart ASA.
+>   - `name` and `unit_name` MAY be set to `SMART-ASA` and `S-ASA`, to
+>     denote that this ASA is Smart and has a controlling application.
+> - Persist the `total`, `decimals`, `default_frozen`, etc. fields for later
+>   use/retrieval.
+> - Return the ID of the created ASA.
+>
+> It is RECOMMENDED for calls to this method to be permissioned, e.g. to only
+> approve transactions issued by the controlling Smart Contract creator.
+
+#### Asset Configuration
+
+```json
+{
+  "name": "asset_config",
+  "args": [
+    { "type": "asset", "name": "config_asset" },
+    { "type": "uint64", "name": "total" },
+    { "type": "uint32", "name": "decimals" },
+    { "type": "bool", "name": "default_frozen" },
+    { "type": "string", "name": "unit_name" },
+    { "type": "string", "name": "name" },
+    { "type": "string", "name": "url" },
+    { "type": "byte[]", "name": "metadata_hash" },
+    { "type": "address", "name": "manager_addr" },
+    { "type": "address", "name": "reserve_addr" },
+    { "type": "address", "name": "freeze_addr" },
+    { "type": "address", "name": "clawback_addr" }
+  ],
+  "returns": { "type": "void" }
+}
+```
+
+Calling `asset_config` configures an existing Smart ASA.
+
+> Upon a call to `asset_config`, a reference implementation SHOULD:
+>
+> - Fail if `config_asset` does not correspond to an ASA controlled by this smart
```

Comment:
Presumably this is done by checking clawback is set to the address of the current app? Should this explicitly be mentioned?

Reply:
I would consider this an implementation detail: your suggestion works, but you could even hardcode the list of assets you want to manage in the TEAL, for instance.

Reply:
Yeah, that's fair. Is it worth listing a couple of them as examples or as a `MAY`?

Reply:
I'd prefer keeping this as light as possible, but I am open to discussion.

What I'd like to stand from the prose is "if the controlling SC created / configured the ASA, then it controls it". Also, this means that when creating it or configuring it, it should ensure control and right syntax/semantics.

If this doesn't show from the prose, then yes, we need to add examples. :)